### ISSUE TITLE
Shared informer has a race condition

### ISSUE THREAD

**Issue by krousey** [kubernetes/kubernetes#59822]
**Is this a BUG REPORT or FEATURE REQUEST?**: Bug

 /kind bug

**What happened**: I was testing a cluster-api machine controller, and I noticed a double channel close panic on test cleanup.

**What you expected to happen**: No panic.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a new shared informer.
2. Concurrently start and add a listener.

There is a path where it can call `run` and `pop` on the `sharedProcessor` twice. When this happens, closing the stop channel will trigger a double close inside of `pop`. Also, this means that 2 copies of `run` are happening on the `sharedProcessor` which seems to violate the guarantee that events to a single handler will be delivered sequentially. Aside from the double close panic on shutdown, this could be the source of heisenbugs in controllers relying on this guarantee.

I'll have a PR out shortly.

cc @kubernetes/sig-api-machinery-bugs

**Comment by krousey** [comment]
Sample crash:

```console
I0213 16:49:17.042] 	panic: close of closed channel
I0213 16:49:17.042] 
I0213 16:49:17.042] goroutine 227 [running]:
I0213 16:49:17.042] k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
I0213 16:49:17.043] 	/go/src/k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x111
I0213 16:49:17.043] panic(0x18e9f00, 0x3ecc6f0)
I0213 16:49:17.043] 	/usr/local/go/src/runtime/panic.go:491 +0x283
I0213 16:49:17.044] k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/client-go/tools/cache.(*processorListener).pop(0xc420713180)
I0213 16:49:17.044] 	/go/src/k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/client-go/tools/cache/shared_informer.go:524 +0x41c
I0213 16:49:17.045] k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/client-go/tools/cache.(*processorListener).(k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/client-go/tools/cache.pop)-fm()
I0213 16:49:17.045] 	/go/src/k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/client-go/tools/cache/shared_informer.go:382 +0x2a
I0213 16:49:17.045] k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc4200c5018, 0xc4200add80)
I0213 16:49:17.046] 	/go/src/k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
I0213 16:49:17.046] created by k8s.io/kube-deploy/ext-apiserver/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
```

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kube-deploy-test/89

**Comment by krousey** [comment]
I have a small test case that has failed 76 out of 31,000 runs. Seems to happen more often with more complicated code to allow for context switches. But it happens quickly enough using the stress command that I can verify it this way. (31000 runs took like than 20 seconds).

**Comment by wild-endeavor** [comment]
@krousey found this fix helpful thank you - could you explain why it happened though?  I don't fully understand where it can call run and pop on sharedprocessor twice... how does that happen?

**Comment by slimhazard** [comment]
Is there a workaround for this issue, if I'm using a version of client-go in which the fix is not implemented?

I'm writing controller code that is obliged to run in a productive cluster at k8s version 1.9.7. The compatible client-go version is 6.0.0, if I want the "checkmark" level in the [compatibility matrix](https://github.com/kubernetes/client-go#compatibility-matrix). I'm reliably getting the double channel close panic from the ``processorListener.pop()`` call in ``shared_informer.go``.

If I'm seeing it right, the fix is available as of v10.0.0. That gets me into ``+-`` on the compatibility matrix, and it seems like a very large jump to take for one issue at shutdown time, whereas everything else is working well with v6.0.0.

So I'd really like to stay at v6.0.0 and implement a workaround. (Of course a backport of the fix wouldn't be so bad either.)

**Comment by slimhazard** [comment]
Update on my own question, in case anyone's reading.

Clicking around the github issues gave me impression that the fix may have been backported to more recent 1.9.x versions of client-go. So I tried building the client against client-go at tag ``kubernetes-1.9.8``, rather than ``v6.0.0``. (This had the curious effect of setting client-go to ``v6.0.1`` in ``go.mod``, although I don't see any such tag with that name in the client-go repo.)

The result is apparently that the client is affected by the bug intermittently, rather than every time. I see the problem at shutdown time (shutting down queues that are fed by event handlers created from shared informers), as the double channel close panic. As indicated above, the panic trace shows the ``processorListener.pop()`` in ``shared_informer.go``. With ``v6.0.0``, the panic happened on every shutdown. With ``kubernetes-1.9.8``/``v6.0.1``, shutdowns are sometimes okay, but sometimes end in the panic.

This is in line with the original report of a race condition that is observed intermittently. I assume that in the runs that end in the panic, sequential delivery was also not guaranteed.

I'm starting to reach the reluctant conclusion that there is no way for a client using shared informers to support kubernetes 1.9.x, unless it either has to live with this intermittent issue in client-go, or use a version of client-go that is not at the "checkmark" level in the compatibility matrix. Would that be right?
